File: 4-algorithms-on-strings/week-1/3_trie_matching_extended.js
const readline = require('readline');


let lineNo = 0;
let text = '';
let n = 0;
const patterns = [];


class Node {
	constructor(key) {
		this.key = key;
		this.edges = [];
		this.prefix = false;
	}
}


class Edge {
	constructor(label, next_node) {
		this.label = label;
		this.next = next_node;
	}
}


const constructTrie = function () {
	let trie = [new Node(0)];
	let node;
	for (let pattern of patterns) {
		node = trie[0];
		for (let char of pattern) {
			const match = node.edges.find(edge => edge.label === char);
			if (match === undefined) {
				const new_node = new Node(trie.length);
				trie.push(new_node);
				const new_edge = new Edge(char, new_node);
				node.edges.push(new_edge);
				node = new_edge.next;
			}
			else {
				node = match.next;
			}
		}
		node.prefix = true;
	}
	trieMatching(trie);
};


const prefixTrieMatching = function (string, trie) {
	const string_match = string;
	let symbol = string_match.shift();
	let v = trie[0];
	while (v.edges.length > 0 && !v.prefix) {
		const match = v.edges.find(edge => edge.label === symbol);
		if (match !== undefined) {
			symbol = string_match.shift();
			v = match.next;
		}
		else
			return false;
	}
	return true;
};


const trieMatching = function (trie) {
	const string = text.split('');
	const runs = string.length;
	const matches = [];
	for (let i = 0; i < runs; i++) {
		const match = prefixTrieMatching(string.slice(i), trie);
		if (match)
			matches.push(i);
	}
	console.log(...matches);
	process.exit();
};


const readInput = function (line) {
	lineNo++;
	if (lineNo === 1) {
		text = line.trim();
	}
	else if (lineNo === 2) {
		n = Number.parseInt(line.trim(), 10);
	}
	else {
		const pattern = line.trim();
		patterns.push(pattern);
		if (patterns.length === n) {
			constructTrie();
		}
	}
};


const rl = readline.createInterface({
	input: process.stdin,
	output: process.stdout,
	terminal: false
});


rl.on('line', readInput);

File: 4-algorithms-on-strings/week-1/3_trie_matching_extended.py
# python3


class Node:
    def __init__(self, key):
        self.key = key
        self.prefix = False
        self.edges = []


class Edge:
    def __init__(self, label, next_node):
        self.label = label
        self.next = next_node


def find_label(edges, label):
    for edge in edges:
        if edge.label == label:
            yield edge


def construct_trie(patterns):
    trie = [Node(0)]
    for pattern in patterns:
        node = trie[0]
        for char in pattern:
            match = next((edge for edge in node.edges if edge.label == char), None)
            if match is None:
                new_node = Node(len(trie))
                trie.append(new_node)
                new_edge = Edge(char, new_node)
                node.edges.append(new_edge)
                node = new_edge.next
            else:
                node = match.next
        node.prefix = True
    return trie


def prefix_trie_matching(text, trie):
    text = list(text)
    symbol = text.pop(0)
    v = trie[0]
    spelled_pattern = ''
    while True:
        if len(v.edges) == 0 or v.prefix:
            return True
        match = next((edge for edge in v.edges if edge.label == symbol), None)
        if match is not None:
            if len(text) > 0:
                symbol = text.pop(0)
            else:
                symbol = None
            v = match.next
            spelled_pattern = spelled_pattern + match[0].label
        else:
            return False


def trie_matching(text, trie):
    for i in range(len(text)):
        match = prefix_trie_matching(text, trie)
        if match:
            print(i, end=' ')
        text = text[1:]


def main():
    text = input()
    n = int(input())
    patterns = []
    for _ in range(n):
        patterns.append(input())
    trie = construct_trie(patterns)
    trie_matching(text, trie)


if __name__ == '__main__':
    main()
